### PR TITLE
[SPARK-13213][SQL] Use CartisianProduct instead of BroacastNestedLoopJoin if possibly

### PR DESCRIPTION
**Pull request by viirya**
JIRA: https://issues.apache.org/jira/browse/SPARK-13213
## What changes were proposed in this pull request?

This patch uses `CartisianProduct` instead of `BroacastNestedLoopJoin` for a logical.join if possibly, as the performance of `CartisianProduct` is improved recently.
## How was the this patch tested?

A benchmark test and its result are added into `BenchmarkWholeStageCodegen`.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala [hunk 1]
```diff
@@ -380,7 +380,12 @@ object ColumnPruning extends Rule[LogicalPlan] {
   /** Applies a projection only when the child is producing unnecessary attributes */
   private def prunedChild(c: LogicalPlan, allReferences: AttributeSet) =
     if ((c.outputSet -- allReferences.filter(c.outputSet.contains)).nonEmpty) {
-      Project(allReferences.filter(c.outputSet.contains).toSeq, c)
+      c match {
+        case BroadcastHint(p) =>
+          BroadcastHint(Project(allReferences.filter(c.outputSet.contains).toSeq, p))
+        case _ =>
+          Project(allReferences.filter(c.outputSet.contains).toSeq, c)
+      }
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala]: This issue is proposed to fix in #11260.

**Comment by viirya** [comment]
retest this please.

**Comment by davies** [comment]
I had created https://github.com/apache/spark/pull/11328 to improve BroadcastNestedLoopJoin, so we should still choose it over CartesianProduct for some cases.

**Comment by viirya** [comment]
Let me close this now.

**Pull request closed without merging**